Added kth roots to Permutation

GermainPoullot opened the pull request:
### 📚 Description

Added 3 functions that deal with k-th roots of permutations:
1) Permutation.kth_roots(k) compute a iterator of over all k-th roots of self.
2) Permutation.has_kth_root(k) determines if self has a k-th root (or several).
3) Permutation.number_of_kth_roots(k) returns the number of k-th roots of self.

Added integer_partition_with_given_parts(n, parts) in partition.py (for use in k-th roots computations):
it creates a iterator over all partitions of n which parts are in the variable parts.

<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->


<!-- Describe your changes here in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Closes #1337" -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [ ] I have linked an issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### ⌛ Dependencies
<!-- List all open pull requests that this PR logically depends on -->
<!--
- #xyz: short description why this is a dependency
- #abc: ...
-->

GermainPoullot committed:
Added kth roots to Permutation

Added integer_partition_with_given_parts in partition.py (for use in kth roots computations).

mantepse commented:
* Integer partitions with given parts already exists:
```
sage: list(Partitions(10, parts_in=[2,3,5]))
[[5, 5], [5, 3, 2], [3, 3, 2, 2], [2, 2, 2, 2, 2]]
```
* I think `Permutation.roots` would be easier to discover.
* The number of k-th roots only depends on the cycle type of the partition.  This should be stated in the docstring - but perhaps, the method should in fact be a method of `Partition`?  I am not sure about this.  You may also want to compare with https://www.findstat.org/StatisticsDatabase/St000811.

mantepse commented:
Further minor things:

* `indexes` should be `indices`
* @fchapoton, there seems to be something wrong with the linter checks: there is almost never a whitespace after the comma.

GermainPoullot committed:
Deleted integer_partitions_with_given_parts

Was redundant.
Added comas.

GermainPoullot committed:
"len" was lacking

GermainPoullot commented:
> * Integer partitions with given parts already exists:

Indeed! Thanks. How did I miss that...
I deleted my function and used this one instead.

>     * I think `Permutation.roots` would be easier to discover.

I agree (especially because .kth_roots() compute by default the square roots), but if someone wants to extend this to Weyl groups or Coxeter groups, "roots" would feel weird I think.

>     * The number of k-th roots only depends on the cycle type of the partition.

You're right. Even thought I think "number_of_kth_roots" is better fit in this module, with its sibblings. I've added the info in doc. 

(Changed the name of `indices` and added spaces after coma.)

GermainPoullot committed:
Improved doc: number roots depend only cyclic type

mantepse commented:
@fchapoton, did you see my comment?  I think it is important, because this might create a huge mess in the near future.

mantepse commented:
> > * I think `Permutation.roots` would be easier to discover.
> 
> I agree (especially because .kth_roots() compute by default the square roots), but if someone wants to extend this to Weyl groups or Coxeter groups, "roots" would feel weird I think.

Excellent point.  I still hope for a better name, but maybe there is none.

> > * The number of k-th roots only depends on the cycle type of the partition.

> You're right. Even thought I think "number_of_kth_roots" is better fit in this module, with its sibblings. I've added the info in doc.

* to get the dict of multiplicities of cycle lengths, you can use `pi.cycle_type().to_exp_dict()`
* I have not checked your algorithm, but if you need key and value of a dict (as in `Cycles`), you should iterate over `Cycles.items()`.
* the inner loop can be replaced by `prod(factorial(x-1) * m**(x-1) for x in partition)`, which is much easier to read.
* I don't see why you first collect the counts into a dict and then take the product.
* variable names should follow the style in https://doc.sagemath.org/html/en/developer/coding_basics.html#python-code-style, in particular https://peps.python.org/pep-0008/
* please remove `partitions.py` from the diff

tscrim commented:
I feel we should call this `nth_roots()` to match what, e.g., field elements do (they have a `nth_root()` method). Although calling it `roots()` is probably the most natural, but as mentioned, it directly conflicts with Coxeter group notions.

There should not be a default for `k`. (What makes 2 so special anyways?)

Remove also the change (removing a line) to `partition.py`.

Add  tests for corner cases of `n=0,1`.

Make sure pyflakes is run. I think `P = Permutations(self.size())` is unused in the `number_of_kth_roots`.

------------------
Some ways you can get more performance:

`itertools` is also a great friend.

This whole part seems to be creating a lot of unnecessary objects:
```
            b = False
            for partition in Partitions(N, parts_in=parts):
                b = True
                count = SetPartitions(N, partition).cardinality()
                for x in partition:
                    count *= factorial(x-1) * m**(x-1)
                Counts[m] += count
            if not b:
                return 0
```
I would refactor out the cardinality method into a standalone function. Likewise, you can just use the `Partitions(N, parts_in=parts)._other_iterator()`, which yields lists. Then with a slight modification, the `cardinality()` could handle input as a list (instead of a `Partition`) by avoiding `to_exp_dict()` and just count the multiplicities directly.

GermainPoullot committed:
Restore partition

GermainPoullot committed:
Change to nth_roots and improve code

Change kth to nth
Improve number_of_nth_roots (better way to iter and more readable)
Idem for has_nth_root
Minor improvements in nth_roots

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5270,49 +5270,52 @@ def shifted_shuffle(self, other):
+        Return all n-th roots of ``self`` (as a generator).
         
-        A k-th root of the permutation self is a permutation Gamma such that Gamma^k == self.
+        A n-th root of the permutation ``self`` is a permutation `\gamma` such that `\gamma^n == self`.
```
It is now `An n-th root` :-)

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5270,49 +5270,52 @@ def shifted_shuffle(self, other):
-        INPUT:
-        
-        - k -- optional integer (default 2), at least 1
+        Note that the number of n-th roots only depend on the cyclic type of ``self``.
```
I think it is `cycle type`, rather than `cyclic type`.

GermainPoullot replied:
Indeed. I'm confused by the French notation.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5270,49 +5270,52 @@ def shifted_shuffle(self, other):
-                * :meth:`has_kth_root`
-                * :meth:`number_of_kth_roots`
+                * :meth:`has_nth_root`
+                * :meth:`number_of_nth_roots`
```
You also may want to compare with the functionality provided by `GroupLibGap`, also speed-wise.  For example:

```
sage: G = GroupLibGAP(SymmetricGroup(5)._libgap_())
sage: G.one().nth_roots(3)
[1,
 (1,2,3),
 (2,3,5),
 (1,5,3),
 (1,2,5),
 (2,5,3),
 (2,3,4),
 (1,3,5),
 (1,4,3),
 (1,5,2),
 (1,2,4),
 (1,3,2),
 (2,4,3),
 (3,4,5),
 (2,5,4),
 (1,3,4),
 (3,5,4),
 (1,4,5),
 (1,4,2),
 (2,4,5),
 (1,5,4)]
```

GermainPoullot replied:
For identity permutation of size 10, computing cubic roots (i.e. `Permutations(10).identity().nth_roots(3)`) takes 5.05 micro on my PC, versus 183 ms with GAP. We gain a factor 1000.
How do I indicate this in the docstring? (More generaly, how and when do we put speed test in doc?)

mantepse replied:
There is no formalized way to do it.  Personally, I like to write a test in the TESTS section which is just fast enough not to be annoying when running the tests (about half a second or so, and so that using the old method (if there is any) one would notice that the test takes too long.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5322,7 +5325,7 @@ def kth_roots(self, k=2):
         def merging_cycles(list_of_cycles):
             """
-            Generate all l-cycles such that its k-th power is the product of cycles in Cycles (which conctains gcd(l, k) cycles of lenght l/gcd(l, k))
+            Generate all l-cycles such that its n-th power is the product of cycles in Cycles (which conctains gcd(l, n) cycles of lenght l/gcd(l, n))
```
This docstring is a bit unclear.  In particular, what is `Cycles`?

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5339,18 +5342,18 @@ def merging_cycles(list_of_cycles):
+        def rewind(L, n):
             """
-            Construct the list M such that M[(j*k)%(len(M))] == L[j].
+            Construct the list M such that M[(j*n)%(len(M))] == L[j].
```
Maybe this docstring could be rewritten without code, but this is really minor.  If not, also enclose it in double single quotes, and stick to the spacing rules.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5339,18 +5342,18 @@ def merging_cycles(list_of_cycles):
-            Construct the list M such that M[(j*k)%(len(M))] == L[j].
+            Construct the list M such that M[(j*n)%(len(M))] == L[j].
             """
             M = [0 for _ in L]
```
I think you can do `[0]*len(L)` here.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5339,18 +5342,18 @@ def merging_cycles(list_of_cycles):
             m = len(M)
             for j in range(m):
-                M[(j*k)%m] = L[j]
+                M[(j*n)%m] = L[j]
```
The linter will require this to be written as `M[(j * k) % m]`.

tscrim replied:
I don't think so. Plus PEP8 says we can put higher precedence operators without spaces (and `[]` counts as an operator). I don't think the linter is (nor should be) that strict.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5362,18 +5365,19 @@ def rewind(L, k):
         
-        # for each length m, collects all product of cycles which k-th power gives the product prod(Cycles[l])
+        # for each length m, collects all product of cycles which n-th power gives the product prod(Cycles[l])
         Possibilities = {m: [] for m in Cycles}
```
I don't think you need a dict here, since you prepare the entries for each cycle separately.  Also, variable names are usually lowercase.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5382,154 +5386,139 @@ def rewind(L, k):
+        for m, N in Cycles.items():
+            N = Integer(N) # I don't know why but ._findfirst doesn't work without
+            parts = [x for x in divisors(n) if gcd(m*x, n) == x]
+            if not Partitions(0, parts_in=[])._findfirst(N, parts):
```
This is a really strange line of code.  `Partitions(0, parts_in=[])` is simply the empty partition.  `_findfirst` is a method without documentation, what is it suppose to do?

GermainPoullot replied:
My goal here is just to know if there exists a partition of `N` with parts in `parts`. I haven't been able to find an efficient method to do it.
Here, I create a parent Partitions_parts_in (writing `Partitions_parts_in(N, parts)` does not seem to work), and then ask for the first element.
`._findfirst` is the method that the class `Partitions_parts_in` uses to iterate over the partitions, if i understand well.
If you have a better idea, I'll take it!

mantepse replied:
`Partitions(N, parts_in=parts).is_empty()`

GermainPoullot replied:
Thanks.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5382,154 +5386,139 @@ def rewind(L, k):
-        P = Permutations(self.size())
+        Cycles = self.cycle_type().to_exp_dict()
+        Result = 1
+        for m, N in Cycles.items():
```
it might be helpful to rename `m` to `length` and `N` to `multiplicity`, or perhaps `m`.  Also, `d` might be a good name for a divisor, instead of `x`.

GermainPoullot replied:
`N` is not a multiplicity, it is the number of cycles of a given length.
I've changed the others.

mantepse replied:
Yes, multiplicty of the cycle length

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5382,154 +5386,139 @@ def rewind(L, k):
+        Result = 1
+        for m, N in Cycles.items():
+            parts = [x for x in divisors(n) if gcd(m*x, n) == x]
+            Result *= sum(SetPartitions(N, pa).cardinality() *
```
It might be helpful to explain this computation in a comment.

videlec replied:
... and not use caml case for variable names, ie `Cycles -> cycles`, `Result -> result`. Caml case is reserved for class names.

mantepse reviewed (commented):
This is beginning to look very good!

fchapoton committed:
Merge branch 'develop' into kth_roots_of_permutations

fchapoton committed:
Update permutation.py

remove some spaces in empty lines

fchapoton committed:
Merge branch 'develop' into kth_roots_of_permutations

fchapoton committed:
Update permutation.py

variables en minuscules

fchapoton committed:
Update permutation.py

some details

fchapoton committed:
Update permutation.py

fchapoton committed:
Update permutation.py

fchapoton committed:
Update permutation.py

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5431,20 +5433,20 @@ def nth_roots(self, n):
             lperm = len(list_of_cycles[0])
             l = lC*lperm
-            perm = [0 for i in range(l)]
+            perm = [0]
```
are you sure?  don't you access perm[j*lC] below?

fchapoton replied:
oops, indeed, sorry

fchapoton committed:
Update permutation.py

less whitespace

fchapoton committed:
Update permutation.py

some fixes

fchapoton committed:
Update permutation.py

less whitespace

fchapoton committed:
Update permutation.py

detail

fchapoton committed:
Update permutation.py

detail

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
+        def merging_cycles(list_of_cycles):
+            """
+            Generate all l-cycles such that its n-th power is the product
+            of cycles in 'cycles' (which contains gcd(l, n) cycles of length l/gcd(l, n))
```
should there be backticks around "l", "n", "gcd(l, n)" and "1/gcd(l, n)"?  Also, possibly change "l" to "k" in the docstring (3-times)

fchapoton replied:
dostrings of inner auxiliary functions are never displayed, except when looking at the code. No need to format them

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
+
+        EXAMPLES::
+
+            sage: Sigma = Permutations(5).identity()
```
"Sigma" might better be lowercase.

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
+        cycles = self.cycle_type().to_exp_dict()
+
+        # for each length m, check if the number of m-cycles can come from a n-th power
+        # (i.e. if you can partitionate m*Cycles[m] into parts of size l with l = m*gcd(l, n))
```
"partitionate" should be "partition".

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
+
+        EXAMPLES::
+
+            sage: Sigma = Permutations(5).identity()
```
"Sigma" might better be lowercase.

fchapoton committed:
Update permutation.py

suggested details

fchapoton committed:
Update permutation.py

fix doctest

mantepse commented on src/sage/combinat/permutation.py:
```diff
@@ -5383,6 +5383,253 @@ def shifted_shuffle(self, other):
+        r"""
+        Decide if ``self`` has n-th roots.
+
+        An n-th root of the permutation ``self`` is a permutation `\gamma` such that `\gamma^n == self`.
```
`\gamma^n == self` doesn't come out properly in the documentation.  Not sure what it should be, possibly
```
``gamma`` such that ``gamma^n == self``
```
or 
```
of the permutation `\sigma` is a permutation `\gamma` such that `\gamma^n = \sigma`.
```
I don't know how to do it properly.

The same comment applies to the docstrings of `n_th_roots` and `number_of_nth_roots`.

I'm sorry :-(

tscrim replied:
> ```
> of the permutation `\sigma` is a permutation `\gamma` such that `\gamma^n = \sigma`.
> ```
This is the one I would do.

fchapoton replied:
done already

fchapoton committed:
Update permutation.py

fixes in doc

mantepse reviewed (approved):
Perfect, thank you!
